### PR TITLE
[v6r14] DataManagement small improvements

### PR DESCRIPTION
**Pull request by phicharp**
* Allow selecting disk replicas in getActiveReplicas() and getReplicas()
* Better check that files are staged in stager, + protect against bad metadata structure
* Minor fix in decorator for executing method with user proxy when processing exceptions (return the actual exception information)
* Use the preferDisk option in the InputData optimizer, the TransformationAgent and in the Interface splitter

### CODE REVIEW

File: DataManagementSystem/Client/DataManager.py [hunk 1]
```diff
@@ -1506,19 +1506,39 @@ def put( self, lfn, fileName, diracSE, path = None ):
   # def putReplica(self,lfn,storageElementName,singleFile=False):
   # def replicateReplica(self,lfn,size,storageElementName,singleFile=False):
 
-  def getActiveReplicas( self, lfns, getUrl = True ):
+  def getActiveReplicas( self, lfns, getUrl = True, diskOnly = False, preferDisk = False ):
```
Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: I don't really se the interest of preferDisk here ? it does not really seem used to me ?

File: WorkloadManagementSystem/Executor/InputData.py [hunk 2]
```diff
@@ -143,13 +143,14 @@ def _resolveInputData( self, jobState, inputData ):
...
+    vo = manifest.getOption( 'VirtualOrganization' )
     startTime = time.time()
     dm = self.__getDataManager( vo )
     if dm is None:
-      return S_ERROR( 'Failed to instantiate DataManager for vo %s' % vo )  
+      return S_ERROR( 'Failed to instantiate DataManager for vo %s' % vo )
     else:
-      result = dm.getActiveReplicas( lfns )  # This will return already active replicas, excluding banned SEs
+      # This will return already active replicas, excluding banned SEs, and removing tape replicas if there are disk replicas
+      result = dm.getActiveReplicas( lfns, preferDisk = True )
```
Comment (phicharp) [inline WorkloadManagementSystem/Executor/InputData.py]: This is the only change, the rest is formatting

File: StorageManagementSystem/Client/StorageManagerClient.py [hunk 3]
```diff
@@ -41,25 +41,27 @@ def getFilesToStage( lfnList ):
...
-        if mDict['Cached']:
+        if 'Cached' not in mDict:
+          failed.setdefault( se, {} )[lfn] = 'No Cached item returned as metadata'
+        elif mDict['Cached']:
           onlineLFNs.add( lfn )
 
   # If the file was found staged, ignore possible errors, but print out errors
+  for se, seFailed in failed.items():
+    gLogger.error( "Errors when getting files metadata", 'at %s' % se )
+    for lfn, reason in seFailed.items():
```
Comment (chaen) [inline StorageManagementSystem/Client/StorageManagerClient.py]: iteritems ?

File: StorageManagementSystem/Client/StorageManagerClient.py [hunk 4]
```diff
@@ -41,25 +41,27 @@ def getFilesToStage( lfnList ):
...
       # is there at least one online?
       for lfn, mDict in fileMetadata['Value']['Successful'].iteritems():
-        if mDict['Cached']:
+        if 'Cached' not in mDict:
+          failed.setdefault( se, {} )[lfn] = 'No Cached item returned as metadata'
+        elif mDict['Cached']:
           onlineLFNs.add( lfn )
 
   # If the file was found staged, ignore possible errors, but print out errors
+  for se, seFailed in failed.items():
```
Comment (chaen) [inline StorageManagementSystem/Client/StorageManagerClient.py]: iteritems ?

**Comment by fstagni** [comment]
It seems OK to me (apart from Chris' minor remarks).

Probably, DataManager would deserve to have its own unit test, and maybe this is a good opportunity to write it.

**Comment by atsareg** [comment]
Review OK, but let Chris review it as well. But why do you put changes like items -> iteritems ?! This is a patch. I am not asking you any more not to mix formatting/beautification into the patches as this will be anyway ignored. Thanks at least for making sometimes comments about what is the real change. But do not make unnecessary code changes which are just a matter of your taste in the patches.

**Comment by chaen** [comment]
The review is OK for me as well, we discussed it before implementation. 

Concerning iteritems, the comment of using this comes from me. The reason behind this is 
* for performance (and it seems like it started to become one of your concern too ;-) )
* for clarity. Now this one deserves a bit more explanations. In python 3, there is no dict.items anymore. When we will have to do the migration, it will really ease our work to know in which case we really need a copy of the elements, and in which one we just wanted to loop over them. It will also help the migration tools such as 2to3. 

Okay, maybe replacing all the items by itermitems in a patch PR is a bit pushing it, but we should adopt this change just like we did for isinstance vs type. 

Concerning beautification. I proposed long ago (and even made) a PR which contains *only* beautification for *all* DIRAC. If we do this, everyone will have to go through the pain of updating once the local repositories, but then, there will not be any beautifying in patch PRs anymore. What do you guys think ?

**Comment by phicharp** [comment]
@Atsareg: sorry again about the beautifying, I hadn't even realized there was any until the branch was uploaded and compared with the head of the branch. I support 200% Chris' proposal to once and hopefully for all get rid of the missing or superfluous blanks (unless some users have automatic features in their editor that is not following the prescriptions that were distributed years ago)

I also thought we had agreed to try and use "features" of python and make the changes as we touch files and not all at once (like replacing type() by is instance(), removing .has_key() and usage of .keys() when they are not necessary in loops. Chris suggested to replace items() by iteritems() for performance reasons in one of my changes, therefore I did it for all in the same file.

**Pull request merged**